pysmt/walkers/generic.py

```diff
@@ -16,12 +16,39 @@
 #   limitations under the License.
 #
 from functools import partial
+from six import with_metaclass
 
 import pysmt.operators as op
 import pysmt.exceptions
 
 
-class Walker(object):
+# NodeType to Function Name
+def nt_to_fun(o):
+    """Returns the name of the walk function for the given nodetype."""
+    return "walk_%s" % op.op_to_str(o).lower()
+
+
+class nodetypes(object):
+    def __init__(self, nodetypes):
+        self.nodetypes = nodetypes
+
+    def __call__(self, func):
+        func.nodetypes = self.nodetypes
+        return func
+
+class MetaNodeTypeHandler(type):
```

Comment:
The current form does not allow chaining of the decorator. We need to test if the attribute is set and extend it if necessary.